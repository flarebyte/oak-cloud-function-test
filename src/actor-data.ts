import { OakAction } from './model';
import { version } from './version';
import { OakFunction } from '../src/model';

const template = {
  version,
  flags: [],
  customStatusDict: {},
  systemFlagsDict: {},
};

export const beforeEachAction: OakAction = {
  ...template,
  name: 'oak-cloud-function-test/action/before-each',
  functionName: 'beforeEachAction',
  description: 'Runs a function before each of the tests is run',
};

export const beforeTestAction: OakAction = {
  ...template,
  name: 'oak-cloud-function-test/action/before-test',
  functionName: 'beforeTestAction',
  description: 'Runs a function before this test is run',
};

export const testAction: OakAction = {
  ...template,
  name: 'oak-cloud-function-test/action/test',
  functionName: 'testAction',
  description: 'Runs as part of the core test',
};

export const beforeEachFunc: OakFunction = {
  ...template,
  name: 'oak-cloud-function-test/func/before-each',
  functionName: 'beforeEachFunc',
  description: 'Runs a function before each of the tests is run',
};

export const beforeTestFunc: OakFunction = {
  ...template,
  name: 'oak-cloud-function-test/func/before-test',
  functionName: 'beforeTestFunc',
  description: 'Runs a function before this test is run',
};

export const testFunc: OakFunction = {
  ...template,
  name: 'oak-cloud-function-test/func/test',
  functionName: 'testFunc',
  description: 'Runs as part of the core test',
};

export const actorData = {
  actionDict: {
    beforeEach: beforeEachAction,
    beforeTest: beforeTestAction,
    test: testAction,
  },

  funcDict: {
    beforeEach: beforeEachFunc,
    beforeTest: beforeTestFunc,
    test: testFunc,
  },
};
