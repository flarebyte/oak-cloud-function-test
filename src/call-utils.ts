const senderSeparator = ' -> ';
const noEmptyString = (value: string): boolean => value.length > 0;

export const joinSenders = (senders: string[]): string =>
  senders.filter(noEmptyString).join(senderSeparator);
export const splitSenders = (sender: string): string[] =>
  sender.split(senderSeparator).filter(noEmptyString);

export const addChildToSender = (parent: string, child: string): string =>
  joinSenders([parent, child]);
