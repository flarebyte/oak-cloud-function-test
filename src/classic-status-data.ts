import { OakStatus } from './model';

const status_flags = {
  success: 'success',
  failure: 'failure',
  invalid: 'invalid'
}

const ok: OakStatus = {
  name: 'ok',
  description: 'The request has succeeded',
  flags: [status_flags.success],
};

const created: OakStatus = {
  name: 'created',
  description: 'The request has been fulfilled and has resulted in one or more new resources being created',
  flags: [status_flags.success],
};

const accepted: OakStatus = {
  name: 'accepted',
  description: 'The request has been accepted for processing but has not yet completed',
  flags: [status_flags.success],
};

const notFound: OakStatus = {
  name: 'not-found',
  description: 'The document was not found',
  flags: [status_flags.invalid],
};
const badRequest: OakStatus = {
  name: 'bad-request',
  description: 'Malformed request syntax',
  flags: [status_flags.invalid],
};

const unauthorized: OakStatus = {
  name: 'unauthorized',
  description: 'Invalid authentication credentials',
  flags: [status_flags.invalid],
};

const forbidden: OakStatus = {
  name: 'forbidden',
  description: 'Insufficient access rights',
  flags: [status_flags.invalid],
};

const payloadTooLarge: OakStatus = {
  name: 'payload-too-large',
  description: 'The request payload is too large',
  flags: [status_flags.invalid],
};

const unsupportedMediaType: OakStatus = {
  name: 'unsupported-media-type',
  description: 'The payload is not in the right format',
  flags: [status_flags.invalid],
};

const internalServiceError: OakStatus = {
  name: 'internal-service-error',
  description: 'Encountered an unexpected condition',
  flags: [status_flags.failure],
};

const serviceUnavailable: OakStatus = {
  name: 'service-unavailable',
  description: 'Unable to handle the request due to a temporary overload or scheduled maintenance',
  flags: [status_flags.failure],
};

const classicStatusDict = {
  ok,
  created,
  accepted,
  notFound,
  badRequest,
  unauthorized,
  forbidden,
  payloadTooLarge,
  unsupportedMediaType,
  internalServiceError,
  serviceUnavailable

}

const classicStatusList: OakStatus[] = Object.values(classicStatusDict);

export { classicStatusDict, classicStatusList };
