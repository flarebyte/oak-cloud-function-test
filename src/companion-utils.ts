import {
  OakAction,
  OakActionCall,
  OakActionCallWrapper,
  OakActionCompanion,
  OakActionCtxCall,
  OakActionRequestEvent,
  OakActionToCall,
  OakCall,
  OakCallWrapper,
  OakCtxCall,
  OakEngineContext,
  OakFunctionCompanion,
  OakRequestEvent,
  OakServiceOperation,
  OakServiceOpToCall,
} from './model';

const mergeTwoActionCompanion = (
  a: OakActionCompanion,
  b: OakActionCompanion
): OakActionCompanion => ({
  explicitCall: { ...a.explicitCall, ...b.explicitCall },
  call: { ...a.call, ...b.call },
  callServiceOperationDict: {
    ...a.callServiceOperationDict,
    ...b.callServiceOperationDict,
  },
});

export const mergeActionCompanions = (
  companions: OakActionCompanion[]
): OakActionCompanion => companions.reduce(mergeTwoActionCompanion);

type OakServiceOperationDict = {
  [name: string]: OakServiceOperation;
};

const wrapCall =
  (serviceOpDict: OakServiceOperationDict, wrapper: OakCallWrapper) =>
  (keyCall: [string, OakCall]): [string, OakCtxCall] =>
    [keyCall[0], wrapper(serviceOpDict[keyCall[0]], keyCall[1])];

export const transformActionCompanion =
  (wrapper: OakCallWrapper) =>
  (companion: OakActionCompanion): OakActionCompanion => {
    const explicitCall = companion.explicitCall;
    const call = Object.fromEntries(
      Object.entries(explicitCall).map(
        wrapCall(companion.callServiceOperationDict, wrapper)
      )
    );

    return {
      explicitCall,
      call,
      callServiceOperationDict: companion.callServiceOperationDict,
    };
  };

export const buildActionCompanion = (
  mapping: OakServiceOpToCall[]
): OakActionCompanion => {
  const explicitCall = Object.fromEntries(
    mapping.map((m) => [m.so.functionName, m.call])
  );
  const tempCtx: OakEngineContext = {
    transactions: [],
    actionTransactions: [],
    systemFlags: [],
    businessOperationFlags: {},
    actionFlags: {},
    functionFlags: {},
  };
  const call = Object.fromEntries(
    mapping.map((m) => [
      m.so.functionName,
      (req: OakRequestEvent) => m.call(tempCtx, req),
    ])
  );
  const callServiceOperationDict = Object.fromEntries(
    mapping.map((m) => [m.so.functionName, m.so])
  );
  return {
    explicitCall,
    call,
    callServiceOperationDict,
  };
};

const mergeTwoFunctionCompanion = (
  a: OakFunctionCompanion,
  b: OakFunctionCompanion
): OakFunctionCompanion => ({
  explicitCall: { ...a.explicitCall, ...b.explicitCall },
  call: { ...a.call, ...b.call },
  actionDict: { ...a.actionDict, ...b.actionDict },
});

export const mergeFunctionCompanions = (
  companions: OakFunctionCompanion[]
): OakFunctionCompanion => companions.reduce(mergeTwoFunctionCompanion);

type OakActionDict = {
  [name: string]: OakAction;
};

const wrapActionCall =
  (actionDict: OakActionDict, wrapper: OakActionCallWrapper) =>
  (keyCall: [string, OakActionCall]): [string, OakActionCtxCall] =>
    [keyCall[0], wrapper(actionDict[keyCall[0]], keyCall[1])];

export const transformFunctionCompanion =
  (wrapper: OakActionCallWrapper) =>
  (companion: OakFunctionCompanion): OakFunctionCompanion => {
    const explicitCall = companion.explicitCall;
    const call = Object.fromEntries(
      Object.entries(explicitCall).map(
        wrapActionCall(companion.actionDict, wrapper)
      )
    );

    return {
      explicitCall,
      call,
      actionDict: companion.actionDict,
    };
  };

export const buildFunctionCompanion =
  (mapping: OakActionToCall[]) =>
  (actionCompanion: OakActionCompanion): OakFunctionCompanion => {
    const explicitCall = Object.fromEntries(
      mapping.map((m) => [m.action.functionName, m.call])
    );
    const tempCtx: OakEngineContext = {
      transactions: [],
      actionTransactions: [],
      systemFlags: [],
      businessOperationFlags: {},
      actionFlags: {},
      functionFlags: {},
    };
    const call = Object.fromEntries(
      mapping.map((m) => [
        m.action.functionName,
        (req: OakActionRequestEvent) => m.call(tempCtx, actionCompanion, req),
      ])
    );
    const actionDict = Object.fromEntries(
      mapping.map((m) => [m.action.functionName, m.action])
    );
    return {
      explicitCall,
      call,
      actionDict,
    };
  };

export type OakFunctionCompanionBuilder = (
  companion: OakActionCompanion
) => OakFunctionCompanion;
