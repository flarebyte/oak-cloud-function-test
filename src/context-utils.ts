import { OakEngineContext } from './model';

export const createEmptyContext = (): OakEngineContext => ({
  transactions: [],
  actionTransactions: [],
  systemFlags: [],
  businessOperationFlags: {},
  actionFlags: {},
  functionFlags: {},
});
