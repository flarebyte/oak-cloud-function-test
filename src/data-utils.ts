import { OakServiceOperation } from './model';

type OakServiceOpObj = { [name: string]: OakServiceOperation };

const mergeTwoServiceOpObj = (
  a: OakServiceOpObj,
  b: OakServiceOpObj
): OakServiceOpObj => ({ ...a, ...b });

export const mergeServiceOperationObjects = (serviceOps: OakServiceOpObj[]) =>
  serviceOps.reduce(mergeTwoServiceOpObj);
