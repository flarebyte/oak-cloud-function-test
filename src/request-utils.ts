import { OakAction, OakBusinessOperation, OakRequestEvent } from './model';

interface OakRequestEventSimpleParams<V, P> {
  businessOperation: OakBusinessOperation;
  comment: string;
  serviceParams: V;
  payload: P;
  callerAction: OakAction;
}

export function createRequestEvent<V extends object, P extends object>({
  businessOperation,
  comment,
  serviceParams,
  payload,
  callerAction,
}: OakRequestEventSimpleParams<V, P>): OakRequestEvent {
  return {
    businessOperation,
    caller: callerAction.name,
    comment,
    serviceParams,
    systemFlags: [],
    payload,
    flags: [],
  };
}
