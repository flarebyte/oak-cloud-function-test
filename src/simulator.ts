import {
  mergeActionCompanions,
  mergeFunctionCompanions,
  OakFunctionCompanionBuilder,
  transformActionCompanion,
  transformFunctionCompanion,
} from './companion-utils';
import {
  OakRequestEvent,
  OakResponseEvent,
  OakActionCompanion,
  OakEngineContext,
  OakCall,
  OakCallWrapper,
  OakServiceOperation,
  OakEventTransaction,
  OakFunctionCompanion,
  OakActionCallWrapper,
  OakAction,
  OakActionRequestEvent,
  OakActionEventTransaction,
  OakActionCall,
} from './model';
import { measureTime } from './perf';
import {
  summarizeActionTransaction,
  summarizeServiceOpTransaction,
  summarizeServiceOpTransactionPerf,
} from './simulator-summarizer';

function cloneValue<A>(value: A): A {
  const jsonStr = JSON.stringify(value);
  const result: A = JSON.parse(jsonStr);
  return result;
}

export class OakSimulator {
  context: OakEngineContext;
  actionCompanion: OakActionCompanion;
  functionCompanion: OakFunctionCompanion;

  constructor() {
    this.context = {
      transactions: [],
      actionTransactions: [],
      systemFlags: [],
    };
    this.actionCompanion = {
      explicitCall: {},
      call: {},
      callServiceOperationDict: {},
    };
    this.functionCompanion = {
      explicitCall: {},
      call: {},
      actionDict: {},
    };
  }

  getCall() {
    return this.actionCompanion.call;
  }
  getActionCall() {
    return this.functionCompanion.call;
  }

  reset() {
    this.context = {
      transactions: [],
      actionTransactions: [],
      systemFlags: [],
    };
  }

  _addTransaction(
    serviceOperation: OakServiceOperation,
    request: OakRequestEvent,
    response: OakResponseEvent,
    nanoSeconds: string
  ) {
    const transaction: OakEventTransaction = {
      id: this.context.transactions.length,
      serviceOperation,
      request,
      response,
      nanoSeconds,
    };
    this.context.transactions.push(cloneValue(transaction));
  }

  _addActionTransaction(
    action: OakAction,
    request: OakActionRequestEvent,
    response: OakResponseEvent
  ) {
    const transaction: OakActionEventTransaction = {
      id: this.context.actionTransactions.length,
      action,
      request,
      response,
    };
    this.context.actionTransactions.push(cloneValue(transaction));
  }

  registerActionCompanions(companions: OakActionCompanion[]) {
    const oneCompanion = mergeActionCompanions(companions);
    const wrapper: OakCallWrapper = (
      serviceOperation: OakServiceOperation,
      wrapped: OakCall
    ) => async (request: OakRequestEvent): Promise<OakResponseEvent> => {
      const thisContext = this.context;
      const reqEvent: OakRequestEvent = {
        ...request,
        systemFlags: [...request.systemFlags, ...thisContext.systemFlags],
      };
      const measuring = measureTime();
      const respEvent = await wrapped(thisContext, reqEvent);
      const nanoSeconds = measuring();
      this._addTransaction(serviceOperation, reqEvent, respEvent, nanoSeconds);
      return respEvent;
    };
    this.actionCompanion = transformActionCompanion(wrapper)(oneCompanion);
  }

  registerFunctionCompanions(companionBuilders: OakFunctionCompanionBuilder[]) {
    const companions = companionBuilders.map(builder =>
      builder(this.actionCompanion)
    );
    const oneCompanion = mergeFunctionCompanions(companions);
    const wrapper: OakActionCallWrapper = (
      action: OakAction,
      wrapped: OakActionCall
    ) => async (request: OakActionRequestEvent): Promise<OakResponseEvent> => {
      const thisContext = this.context;
      const reqEvent: OakActionRequestEvent = {
        ...request,
        systemFlags: [...request.systemFlags, ...thisContext.systemFlags],
      };
      const respEvent = await wrapped(
        thisContext,
        this.actionCompanion,
        reqEvent
      );
      this._addActionTransaction(action, reqEvent, respEvent);
      return respEvent;
    };
    this.functionCompanion = transformFunctionCompanion(wrapper)(oneCompanion);
  }

  toFullInfo() {
    return JSON.stringify(this.context.transactions, null, 2);
  }

  toInfo() {
    return JSON.stringify(
      this.context.transactions.map(summarizeServiceOpTransaction),
      null,
      2
    );
  }

  toSimplifiedTx() {
    return JSON.stringify(
      {
        tx: this.context.transactions.map(summarizeServiceOpTransaction),
        actionTx: this.context.actionTransactions.map(
          summarizeActionTransaction
        ),
      },
      null,
      2
    );
  }

  toServiceOpPerf() {
    return this.context.transactions.map(summarizeServiceOpTransactionPerf);
  }
}
