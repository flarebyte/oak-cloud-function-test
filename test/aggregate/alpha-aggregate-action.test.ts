import {
  OakActionRequestEvent,
} from '../../src/model';
import { OakSimulator } from '../../src/simulator';
import { bizOperationObj } from '../fixture-business-operation';
import { s1DevCompanion } from '../storage/s1-storage-dev';
import { createS1Params } from '../storage/s1-storage-factory';
import { aggregateCompanion } from './alpha-aggregate-action';
import { AggregateTask, CityPayload } from './alpha-aggregate-model';
import { createRequestEvent } from '../../src/request-utils';
import { testAction } from '../fixture-action';
import { S1StorageParams } from '../storage/s1-storage-model';

const sumRequest: OakActionRequestEvent = {
  caller: 'test',
  comment: 'sum',
  flags: [],
  params: { task: AggregateTask.Sum },
  systemFlags: [],
  payload: {},
};

describe('Alpha Aggregate', () => {
  const simulator = new OakSimulator();
  simulator.registerActionCompanions([s1DevCompanion]);
  simulator.registerFunctionCompanions([aggregateCompanion]);
  const call = simulator.getCall();
  const callAction = simulator.getActionCall();

  beforeEach(() => {
    simulator.reset();
  });
  it('should show a successful sum', async () => {
    await call.writeS1(
      createRequestEvent<S1StorageParams, CityPayload>({
        businessOperation: bizOperationObj.writeLondonData,
        comment: 'write to city/london',
        serviceParams: createS1Params('city/london'),
        payload: {
          data: [2, 4, 8],
        },
        callerAction: testAction,
      })
    );
    await call.writeS1(
      createRequestEvent<S1StorageParams, CityPayload>({
        businessOperation: bizOperationObj.writeParisData,
        comment: 'write to city/paris',
        serviceParams: createS1Params('city/paris'),
        payload: {
          data: [1, 3, 5],
        },
        callerAction: testAction,
      })
    );
    const resp = await callAction.aggregateData(sumRequest);
    expect(resp.status.name).toEqual('ok');
    expect(simulator.toSimplifiedTx()).toMatchSnapshot();
  });
});
