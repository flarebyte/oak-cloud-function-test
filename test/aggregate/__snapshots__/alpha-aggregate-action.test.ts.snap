// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alpha Aggregate should show a successful sum 1`] = `
"{
  \\"tx\\": [
    {
      \\"id\\": 0,
      \\"businessOperationName\\": \\"write-to-london\\",
      \\"serviceOperationName\\": \\"company:s1:write\\",
      \\"serviceName\\": \\"company:s1\\",
      \\"resourceName\\": \\"s3:london\\",
      \\"caller\\": \\"unit-tests\\",
      \\"serviceParams\\": {
        \\"path\\": \\"city/london\\"
      },
      \\"systemFlags\\": \\"\\",
      \\"reqComment\\": \\"write to city/london\\",
      \\"reqPayload\\": {
        \\"data\\": [
          2,
          4,
          8
        ]
      },
      \\"reqFlags\\": \\"\\",
      \\"statusName\\": \\"ok\\",
      \\"respComment\\": \\"Success\\",
      \\"respPayload\\": {
        \\"message\\": \\"Saved\\"
      },
      \\"respFlags\\": \\"\\"
    },
    {
      \\"id\\": 1,
      \\"businessOperationName\\": \\"write-to-paris\\",
      \\"serviceOperationName\\": \\"company:s1:write\\",
      \\"serviceName\\": \\"company:s1\\",
      \\"resourceName\\": \\"s3:paris\\",
      \\"caller\\": \\"unit-tests\\",
      \\"serviceParams\\": {
        \\"path\\": \\"city/paris\\"
      },
      \\"systemFlags\\": \\"\\",
      \\"reqComment\\": \\"write to city/paris\\",
      \\"reqPayload\\": {
        \\"data\\": [
          1,
          3,
          5
        ]
      },
      \\"reqFlags\\": \\"\\",
      \\"statusName\\": \\"ok\\",
      \\"respComment\\": \\"Success\\",
      \\"respPayload\\": {
        \\"message\\": \\"Saved\\"
      },
      \\"respFlags\\": \\"\\"
    },
    {
      \\"id\\": 2,
      \\"businessOperationName\\": \\"read-from-london\\",
      \\"serviceOperationName\\": \\"company:s1:read\\",
      \\"serviceName\\": \\"company:s1\\",
      \\"resourceName\\": \\"s3:london\\",
      \\"caller\\": \\"action/alpha-aggregate-europe\\",
      \\"serviceParams\\": {
        \\"path\\": \\"city/london\\"
      },
      \\"systemFlags\\": \\"\\",
      \\"reqComment\\": \\"read from London\\",
      \\"reqPayload\\": {},
      \\"reqFlags\\": \\"\\",
      \\"statusName\\": \\"ok\\",
      \\"respComment\\": \\"Success\\",
      \\"respPayload\\": {
        \\"data\\": [
          2,
          4,
          8
        ]
      },
      \\"respFlags\\": \\"\\"
    },
    {
      \\"id\\": 3,
      \\"businessOperationName\\": \\"read-from-paris\\",
      \\"serviceOperationName\\": \\"company:s1:read\\",
      \\"serviceName\\": \\"company:s1\\",
      \\"resourceName\\": \\"s3:paris\\",
      \\"caller\\": \\"action/alpha-aggregate-europe\\",
      \\"serviceParams\\": {
        \\"path\\": \\"city/paris\\"
      },
      \\"systemFlags\\": \\"\\",
      \\"reqComment\\": \\"read from Paris\\",
      \\"reqPayload\\": {},
      \\"reqFlags\\": \\"\\",
      \\"statusName\\": \\"ok\\",
      \\"respComment\\": \\"Success\\",
      \\"respPayload\\": {
        \\"data\\": [
          1,
          3,
          5
        ]
      },
      \\"respFlags\\": \\"\\"
    },
    {
      \\"id\\": 4,
      \\"businessOperationName\\": \\"write-to-europe\\",
      \\"serviceOperationName\\": \\"company:s1:write\\",
      \\"serviceName\\": \\"company:s1\\",
      \\"resourceName\\": \\"s3:europe\\",
      \\"caller\\": \\"action/alpha-aggregate-europe\\",
      \\"serviceParams\\": {
        \\"path\\": \\"city/europe\\"
      },
      \\"systemFlags\\": \\"\\",
      \\"reqComment\\": \\"write to europe\\",
      \\"reqPayload\\": {
        \\"data\\": [
          14,
          9
        ]
      },
      \\"reqFlags\\": \\"\\",
      \\"statusName\\": \\"ok\\",
      \\"respComment\\": \\"Success\\",
      \\"respPayload\\": {
        \\"message\\": \\"Saved\\"
      },
      \\"respFlags\\": \\"\\"
    }
  ],
  \\"actionTx\\": [
    {
      \\"id\\": 0,
      \\"actionName\\": \\"alpha-aggregate-europe\\",
      \\"caller\\": \\"test\\",
      \\"params\\": {
        \\"task\\": 0
      },
      \\"systemFlags\\": \\"\\",
      \\"reqComment\\": \\"sum\\",
      \\"reqPayload\\": {},
      \\"reqFlags\\": \\"\\",
      \\"statusName\\": \\"ok\\",
      \\"respComment\\": \\"Success\\",
      \\"respPayload\\": {
        \\"message\\": \\"Saved\\"
      },
      \\"respFlags\\": \\"\\"
    }
  ]
}"
`;
